interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/QueueSubscriber.scala

```diff
@@ -1,56 +1,77 @@
-import scalaz.zio.stream.Take
-import scalaz.zio.stream.Take.{ End, Fail, Value }
-import scalaz.zio.{ Queue, Runtime, UIO }
-
-private[reactiveStreams] class QueueSubscriber[A](
-  runtime: Runtime[_],
-  q: Queue[Take[Throwable, A]]
-) extends Subscriber[A] {
-
-  // all signals in reactive streams are serialized, so we don't need any synchronization
-  private var subscriptionOpt: Option[Subscription] = None
-  private var signalledDemand                       = 0
-
-  def signalDemand: UIO[Unit] =
-    subscriptionOpt.fold(UIO.unit) { subscription =>
-      q.size.flatMap {
-        case 0 if q.capacity - signalledDemand > 0 => UIO(subscription.request((q.capacity - signalledDemand).toLong))
-        case _                                     => UIO.unit
-      }
+import scalaz.zio.stream.ZStream
+import scalaz.zio.stream.ZStream.Fold
+import scalaz.zio.{ IO, Queue, Runtime, Task, UIO, ZIO }
+
+private[reactiveStreams] object QueueSubscriber {
+  def make[A](capacity: Int): ZIO[Any, Nothing, (Subscriber[A], ZStream[Any, Throwable, A])] =
+    for {
+      runtime <- UIO.runtime[Any]
+      q       <- Queue.bounded[A](capacity)
+      qs      = new QueueSubscriber[A](runtime, q)
+    } yield (qs, qs)
+}
+
+private class QueueSubscriber[A](runtime: Runtime[_], q: Queue[A])
+    extends Subscriber[A]
+    with ZStream[Any, Throwable, A] {
+
+  private val capacity: Long = q.capacity.toLong
+
+  private var completed: Boolean                 = false
+  private var failed: Option[Throwable]          = None
+  private var subscription: Option[Subscription] = None
+
+  override def fold[R1 <: Any, E1 >: Throwable, A1 >: A, S]: Fold[R1, E1, A1, S] =
+    IO.succeedLazy { (s, cont, f) =>
+      def loop(s: S, demand: Long): ZIO[R1, E1, S] =
+        if (!cont(s)) subscription.fold(UIO.unit)(s => UIO(s.cancel())) *> q.shutdown.const(s)
+        else
+          q.size.flatMap { n =>
+            val empty = n <= 0
+            if (empty && completed) q.shutdown.const(s)
+            else if (empty && failed.isDefined) q.shutdown *> Task.fail(failed.get)
+            else if (empty && (demand < q.capacity))
+              subscription.fold(UIO.unit)(s => UIO(s.request(capacity - demand))) *> loop(s, capacity)
+            else q.take.flatMap(f(s, _)).flatMap(loop(_, demand - 1))
+          } <> failed.fold[Task[S]](UIO.succeed(s))(Task.fail)
+      loop(s, capacity)
     }
 
-  override def onSubscribe(subscription: Subscription): Unit = {
-    if (subscription == null) throw new NullPointerException("s was null in onSubscribe")
-    if (subscriptionOpt.isDefined) {
-      subscription.cancel()
-    } else {
-      subscriptionOpt = Some(subscription)
-      runtime.unsafeRunAsync_(q.awaitShutdown *> UIO(subscription.cancel()))
-      subscription.request(q.capacity.toLong)
+  override def onSubscribe(s: Subscription): Unit = {
+    if (s == null) throw new NullPointerException("s was null in onSubscribe")
+    synchronized {
```

Comment:
Feels weird to use synchronized here. It is only a guard against misuse of the Subscriber. `onSubscribe` should only be called once in the lifetime of a `QueueSubscriber`. Synchronizing here saves us a `Promise.await` in each call to `loop` in `fold`.

Reply:
I'd just throw an exception here if `onSubscribe` is somehow called twice and let the fiber die. The graceful handling of that here might mask implementation errors.

Reply:
According to the spec, we must cancel the subscription and must not fail, so that is not possible.

Reply:
I see. Ok then. I'm good with the use of `synchronized` here.